src/main/java/io/jsonwebtoken/impl/DefaultJwtParser.java

```diff
@@ -342,6 +350,14 @@ public Jwt parse(String jwt) throws ExpiredJwtException, MalformedJwtException,
             //re-create the jwt part without the signature.  This is what needs to be signed for verification:
             String jwtWithoutSignature = base64UrlEncodedHeader + SEPARATOR_CHAR + base64UrlEncodedPayload;
 
+            if (signatureAlgorithm != null &&
+                    (signatureAlgorithm.compareTo(algorithm) != 0)) {
+                String message = "The expected algorithm " + signatureAlgorithm.getValue() + " does not match " +
```

Comment:
Let's have this be:

`The 'alg' JWT header value of '" + algorithm.value() + " does not match the parser required value of '" + requiredSignatureAlgorithm.value() + "'.";`

Also the last sentence is unnecessary because we don't want to speculate as to what might have happened and cause undue alarm.